### PR TITLE
Firmware tool

### PR DESCRIPTION
**Pull request by loucass003**
Requires SlimeVR/SolarXR-Protocol#125

### CODE REVIEW

File: .github/CODEOWNERS
Comment (ImUrX) [inline .github/CODEOWNERS:1]: this file got edited because its crlf originally for some reason

File: gui/.eslintrc.json
Comment (ImUrX) [inline gui/.eslintrc.json:1]: w-what happened to eslint

**Review by ImUrX — CHANGES_REQUESTED** [review]

File: gui/.eslintrc.json
Comment (ImUrX) [reply gui/.eslintrc.json:1]: oh it was me nvm

File: gui/src/components/commons/VerticalStepper.tsx [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+
+  const goTo = (id: string) => {
+    const step = steps.findIndex(({ id: stepId }) => stepId === id);
+    if (step === -1) throw new Error('step not found');
+
+    setStep(step);
+  };
+
+  return (
+    <ol className="relative border-l border-gray-700 text-gray-400 ">
```
Comment (ImUrX) [inline gui/src/components/commons/VerticalStepper.tsx:125]:
```suggestion
    <ol className="relative border-l border-gray-700 text-gray-400">
```

**Review by ImUrX — CHANGES_REQUESTED** [review]

File: gui/src/components/firmware-tool/SelectFirmwareStep.tsx [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+  prevStep: () => void;
+  goTo: (id: string) => void;
+}) {
+  const { l10n } = useLocalization();
+  const { selectVersion, newConfig, defaultConfig } = useFirmwareTool();
+  const { isFetching, data: firmwares } = useGetFirmwaresVersions({});
+
+  const { control, watch } = useForm<{ thirdParty: boolean }>({});
+
+  const showTirdParty = watch('thirdParty');
```
Comment (ImUrX) [inline gui/src/components/firmware-tool/SelectFirmwareStep.tsx:27]: `showThirdParty` instead of `showTirdParty`

**Review by ImUrX — CHANGES_REQUESTED** [review]

File: gui/src/components/firmware-tool/BoardPinsStep.tsx [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+                  placeholder={l10n.getString(
+                    'firmware-tool-board-pins-step-battery-type'
+                  )}
+                  direction="up"
+                  display="block"
+                  items={batteryTypes.map((battery) => ({
+                    label: l10n.getString(
+                      'firmware-tool-board-pins-step-battery-type-' + battery
+                    ),
+                    value: battery,
```
[the remark this answers is not in the document]
Comment (loucass003) [reply gui/src/components/firmware-tool/BoardPinsStep.tsx:111]: Yes it is, it is the value of the dropdown

**Review by Eirenliel — CHANGES_REQUESTED** [review]
TL;DR

**Review by ImUrX — CHANGES_REQUESTED** [review]

File: server/desktop/src/main/java/dev/slimevr/desktop/firmware/DesktopSerialFlashingHandler.kt [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+			error("unable to open port")
+		}
+		this.port = comPort
+	}
+
+	override fun closeSerial() {
+		val p = port ?: error("no port to close")
+		try {
+			p.closePort()
+			println("Port closed")
```
Comment (ImUrX) [inline server/desktop/src/main/java/dev/slimevr/desktop/firmware/DesktopSerialFlashingHandler.kt:30]: use `LogManager.info`

**Review by ImUrX — CHANGES_REQUESTED** [review]

**Review by Eirenliel — APPROVED** [review]

File: gui/src/components/firmware-tool/FlashingMethodStep.tsx [hunk 5]
```diff
@@ -0,0 +1,405 @@
...
+          then: (s) =>
+            s
+              .shape({
+                selectedDevices: object(),
+              })
+              .required(),
+          otherwise: (s) => s.optional(),
+        }),
+      }) as ObjectSchema<FlashingMethodForm>
+    ),
```
[the remark this answers is not in the document]
Comment (loucass003) [reply gui/src/components/firmware-tool/FlashingMethodStep.tsx:324]: nope it is not and there is no other way anyway

File: server/core/src/main/java/dev/slimevr/firmware/FirmwareUpdateHandler.kt [hunk 6]
```diff
@@ -0,0 +1,506 @@
...
+
+	fun addListener(channel: FirmwareUpdateListener) {
+		listeners.add(channel)
+	}
+
+	fun removeListener(channel: FirmwareUpdateListener) {
+		listeners.removeIf { channel == it }
+	}
+
+	fun destroy() {
```
[the remark this answers is not in the document]
Comment (ButterscotchV) [reply server/core/src/main/java/dev/slimevr/firmware/FirmwareUpdateHandler.kt:81]: Is that not what `cancelUpdates()` is for? What is this method actually doing?

File: server/core/src/main/java/dev/slimevr/firmware/FirmwareUpdateHandler.kt [hunk 7]
```diff
@@ -0,0 +1,506 @@
...
+			update(FirmwareUpdateStatus.PROVISIONING)
+		}
+
+		if (status == ProvisioningStatus.DONE) {
+			update(FirmwareUpdateStatus.DONE)
+		}
+
+		if (status == ProvisioningStatus.CONNECTION_ERROR || status == ProvisioningStatus.COULD_NOT_FIND_SERVER) {
+			update(FirmwareUpdateStatus.ERROR_PROVISIONING_FAILED)
+		}
```
Comment (ButterscotchV) [inline server/core/src/main/java/dev/slimevr/firmware/FirmwareUpdateHandler.kt:479]: Can probably use a `when (status) {}` block here.

File: server/core/src/main/java/dev/slimevr/firmware/OTAUpdateTask.kt [hunk 8]
```diff
@@ -0,0 +1,180 @@
...
+					deviceId,
+					FirmwareUpdateStatus.REBOOTING,
+				),
+			)
+		}
+	}
+
+	companion object {
+		private const val FLASH = 0
+		private const val PORT = 8266
```
Comment (ButterscotchV) [inline server/core/src/main/java/dev/slimevr/firmware/OTAUpdateTask.kt:176]: Is ESP32 not planned for support here? ESP32 uses port `3232` for OTA but pretty much everything else is the same iirc.
